src/main/java/com/google/edith/QueryItems.java

```diff
@@ -0,0 +1,151 @@
+        String unit = "";
+
+        for (String expirationPiece : expirationNumberAndUnit) {
+          try {
+            number = (double) Double.parseDouble(expirationPiece);
+          } catch (NumberFormatException e) {
+            unit = expirationPiece;
+          }
+        }
+        if (unit.toLowerCase().equals("weeks")) {
+          number = number * 7;
+        } else if (unit.toLowerCase().equals("months")) {
+          number = number * 30;
+        }
+
+        if (number < daysPassedSinceReceipt) {
+          itemsToBuy.add(item);
+        }
+      }
+    }
+
+    Gson gson = new Gson();
+    return gson.toJson(itemsToBuy);
+  }
+
+  /**
+   * Determines the number of days between when the receipt was stored in datastore and the current
+   * time.
+   */
+  private double findTimePassed(String receiptDateString) {
+    Date currentDate = new Date();
+    Date receiptDate;
+    try {
+      receiptDate = new SimpleDateFormat("yyyy-mm-dd").parse(receiptDateString);
+    } catch (ParseException e) {
+      receiptDate = new Date();
+    }
+    long timePassedSinceReceipt = currentDate.getTime() - receiptDate.getTime();
+    double daysPassedSinceReceipt = (timePassedSinceReceipt / (1000 * 60 * 60 * 24)) / 24;
+    return daysPassedSinceReceipt;
+  }
+
+  private List<Receipt> queryReceipts() {
+    String id = userService.getCurrentUser().getUserId();
+    Query query =
+        new Query("Receipt")
+            .setFilter(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, id));
+    PreparedQuery receiptResults = datastore.prepare(query);
+    List<Entity> receiptEntities = receiptResults.asList(FetchOptions.Builder.withLimit(3));
+    List<Receipt> receipts = new ArrayList<>();
+
+    for (Entity receiptEntity : receiptEntities) {
+      Key entityKey = receiptEntity.getKey();
+      Query itemQuery = new Query("Item", entityKey);
+      PreparedQuery results = datastore.prepare(itemQuery);
+      List<Entity> itemEntities = results.asList(FetchOptions.Builder.withLimit(Integer.MAX_VALUE));
+      Item[] items = createItemObjects(itemEntities);
+      String userId = (String) receiptEntity.getProperty("userId");
+      String storeName = (String) receiptEntity.getProperty("storeName");
+      String date = (String) receiptEntity.getProperty("date");
+      String name = (String) receiptEntity.getProperty("name");
+      String fileUrl = (String) receiptEntity.getProperty("fileUrl");
+      float totalPrice = (float) ((double) receiptEntity.getProperty("price"));
+      Receipt receipt = new Receipt(userId, storeName, date, name, fileUrl, totalPrice, items);
+      receipts.add(receipt);
+    }
+    return receipts;
+  }
+
+  /**
+   * Creates an array of Item objects from entites of kind Item found in the datastore.
+   *
+   * @param entities - entities of kind Item found in datastore.
+   * @return Item[] - array of Item objects.
+   */
+  public Item[] createItemObjects(List<Entity> entities) {
```

Comment:
We should always prefer Lists (ideally ImmutableLists) to object arrays

Reply:
the issue here and with the comment below is that the data model takes an array. i could have the function return an immutable list but would have to turn it back into an array to go into datastore